arrows/vxl/color_commonality_filter.cxx

```diff
@@ -0,0 +1,608 @@
+/*ckwg +5
+ * Copyright 2019 by Kitware, Inc. All Rights Reserved. Please refer to
+ * KITWARE_LICENSE.TXT for licensing information, or contact General Counsel,
+ * Kitware, Inc., 28 Corporate Drive, Clifton Park, NY 12065.
+ */
+
+#include "color_commonality_filter.h"
+
+#include <arrows/vxl/image_container.h>
+
+#include <vil/vil_fill.h>
+#include <vil/vil_image_view.h>
+#include <vil/vil_crop.h>
+#include <vil/algo/vil_gauss_filter.h>
+#include <vgl/vgl_box_2d.h>
+#include <vgl/vgl_intersection.h>
+
+#include <limits>
+#include <type_traits>
+
+
+namespace kwiver {
+namespace arrows {
+namespace vxl {
+
+namespace {
+
+/// Settings for the below functions
+struct color_commonality_filter_settings
+{
+  // Resolution per channel when building a histogram detailing commonality
+  unsigned int resolution_per_channel;
+
+  // Scale the output image (which will by default by in the range [0,1])
+  // by this amount. Set as 0 to scale to the input type-specific maximum.
+  unsigned int output_scale_factor;
+
+  // Instead of computing the per-pixel color commonality out of all the
+  // pixels in the entire image, should we instead compute it in grids
+  // windowed across it?
+  bool grid_image;
+  unsigned grid_resolution_height;
+  unsigned grid_resolution_width;
+
+  // [Advanced] A pointer to a temporary buffer for the histogram, which
+  // can (a) prevent having to reallocate it over and over again, and
+  // (b) allow the user to use it as a by-product of the main operation.
+  // If set to NULL, will use an internal histogram buffer.
+  std::vector<unsigned>* histogram;
+
+  // Default constructor
+  color_commonality_filter_settings()
+    : resolution_per_channel(8),
+      output_scale_factor(0),
+      grid_image(false),
+      grid_resolution_height(5),
+      grid_resolution_width(6),
+      histogram(NULL)
+  {}
+};
+
+// Simple helper functions
+inline bool is_power_of_two( const unsigned num )
+{
+  return ( (num > 0) && ((num & (num - 1)) == 0) );
+}
+
+inline unsigned int integer_log2( unsigned int value )
+{
+  unsigned int l = 0;
+  while( (value >> l) > 1 ) ++l;
+  return l;
+}
+
+// Intersect a region with some image boundaries
+template< typename PixType >
+void check_region_boundaries( vgl_box_2d< int >& bbox,
+  const vil_image_view< PixType >& img )
+{
+  vgl_box_2d<int> boundaries( 0, img.ni(), 0, img.nj() );
+  bbox = vgl_intersection( boundaries, bbox );
+}
+
+// Point an image view to a rectangular region in the image
+template< typename PixType, typename BoxType >
+void point_view_to_region( const vil_image_view< PixType >& src,
+                           const vgl_box_2d< BoxType >& region,
+                           vil_image_view< PixType >& dst )
+{
+  // Early exit case, no crop required
+  if( region.min_x() == 0 && region.min_y() == 0 &&
```

Comment:
Instead of equality, should this be if the region is equal to or larger than the input image?